### ISSUE TITLE
FormDataConsumer not working properly inside 'Filter'

### ISSUE THREAD

**Issue by afilp** [marmelab/react-admin#2308]
**What you were expecting:**

`FormDataConsumer` to be used in Filters, so that we can have a filter with `subcat` values filtered based on another filter with `cat` values.

**What happened instead:**
The filter does not appear correctly:

![image](https://user-images.githubusercontent.com/7850073/45394044-63493a00-b637-11e8-80a8-b3e34b47aca1.png)

**Other information:**

```
      <FormDataConsumer>
        {({ formData, ...rest }) =>
          formData.cat_id && (
            <ReferenceInput
              // key={formData.cat_id}
              label="ΥΠΟΚΑΤΗΓΟΡΙΑ" 
              source="subCat_id"
              reference="subCategories"
              filter={{ cat_id: formData.cat_id }}
              {...rest}
            >
              <SelectInput optionText="name" />
            </ReferenceInput>
          )
        }
      </FormDataConsumer>
```

**Environment**

* React-admin version: v2.2

**Comment by afilp** [comment]
UPDATE: I removed the condition and also added a label to <FormDataConsumer> and it shows in the filters. However the `filter` prop has no effect:

```
      <FormDataConsumer>
        {({ formData, ...rest }) =>
          formData.cat_id && (
            <ReferenceInput
              // key={formData.cat_id}
              label="ΥΠΟΚΑΤΗΓΟΡΙΑ" 
              source="subCat_id"
              reference="subCategories"
              filter={{ cat_id: formData.cat_id }}
              {...rest}
            >
              <SelectInput optionText="name" />
            </ReferenceInput>
          )
        }
      </FormDataConsumer>
```

**Comment by djhi** [comment]
`FormDataConsumer` can only work in `TabbedForm` and `SimpleForm` currently as it specifically targets the redux-form named `record-form`. Marking this as an enhancement

**Comment by afilp** [comment]
@djhi That would be great to have as a feature, thanks a lot!

P.S. This can help a lot in a "wizard-like" filtering of the data.

**Comment by fzaninotto** [comment]
I don't know what you mean by "wizard-like" filtering, but we've implemented a filter as tabs in the demo, see

https://github.com/marmelab/react-admin/blob/38d8030b4bdd1fbab51bcda148c34b76a9606b47/examples/demo/src/commands/CommandList.js#L56

**Comment by afilp** [comment]
@fzaninotto, I read the code there (thanks for the suggestion), I believe this is something different.

I want a filter in `filters` whose **appearance** and **values** **depend** on another filter.

For example, I have `categories` and `subCategories` as filters. If the `categories` filter is not selected, the `subCategories` filter disappears (it does not have any effect in the query, etc.). If the `categories` filter is selected, then the `subCategories` filter is available for selection AND (optionally) its values are filtered by the chosen `category`. 

In other words, exactly like what `formDataConsumer` does (for appearance and filtered values of an input field), but inside the `Filters` logic.

This is what I mean `wizard-like`, we can then choose filters based on other filters.

Another example: Car `model` and `brand` (do not show all available brands in the filter, but only show those related with the selected `model` filter value).

There are many other such use cases (Date From/To, etc.).

Thanks!

**Comment by fzaninotto** [comment]
Gotcha.

You can probably do it using a custom wrapper for your input connected to the filter form in redux-form.

I'm not sure we'll implement that in core, because there are tons of corner cases. I believe this should stay in userland.

**Comment by fzaninotto** [comment]
Probably fixed by #2656

**Comment by dennes544** [comment]
Found solution to filter based on another filter without FormDataConsumer

`<ReferenceInput source="filialid" reference="filials">
            <SelectInput optionText="filialname"/>
        </ReferenceInput>

        {props.filterValues.filialid &&
        <ReferenceInput
            key={props.filterValues.filialid}
            source="cabinetid"
            reference="cabinets"
            perPage={200}
            filter={{ filialid: props.filterValues.filialid }}
        >
            <SelectInput optionText="cabinetname"/>
        </ReferenceInput>}`

Works like a charm

**Comment by amayatsky** [comment]
In case anyone cames in here wondering how to reset the dependent filter value, here is a solution:
```
<FormDataConsumer form={'filterForm'} alwaysOn source='stateId'>
      {({ formData, dispatch, ...rest }) => (
        <ReferenceInput
          resource={props.resource}
          source='stateId'
          reference='states'
          onChange={value => {
            dispatch(change('filterForm', 'districtId', null));
          }}
          allowEmpty>
          <SelectInput optionText='name' />
        </ReferenceInput>
      )}
    </FormDataConsumer>

    {(props.filterValues.stateId || props.filterValues.districtId) && (
      <ReferenceInput
        key={props.filterValues.stateId}
        source='districtId'
        reference='districts'
        filter={{ stateId: props.filterValues.stateId }}
        alwaysOn>
        <SelectInput optionText='name' />
      </ReferenceInput>
    )}
```

**Comment by kopax** [comment]
I think I am hitting the same issue when trying to create a depend field (like categories and sub categories.

This is what I have tried

```jsx
    <Create {...props}>
      <SimpleForm>
        <FormDataConsumer>
          {({ formData, ...rest }) => (
            <>
              <ReferenceInput
                validate={[required()]}
                source="companyId"
                reference={references.companyManagement?.companies?.name}
                {...rest}
              >
                <SelectInput
                  optionText="name"
                />
              </ReferenceInput>
              <br />
              <ReferenceInput
                validate={[required()]}
                source="postalAddressId"
                perPage={1000}
                sort={{
                  field: 'city',
                  order: 'ASC',
                }}
                filter={{
                  // companyId: formData.companyId
                  companyId: 1,
                }}
                reference={references.companyManagement?.postalAddresses?.name}
                {...rest}
              >
                <SelectInput
                  optionText="fullAddress"
                />
              </ReferenceInput>
            </>
          )}
        </FormDataConsumer>
      </SimpleForm>
    </Create>
```


The value passed in `CRUD_GET_MANY_ACCUMULAT` is:

     "{"resource":"company-management/companies","relatedTo":"promotion-management/promotions@companyId","pagination":{"page":1,"perPage":25},"sort":{"field":"id","order":"DESC"},"filter":{}}"

For some reason, `filter` is always empty, I don't get how is set `props.filterValues` in your code sample, do you mind giving a bit more explanation ?
